### PR TITLE
FramePack: inference and training support

### PR DESCRIPTION
**Pull request by kohya-ss**
Add FramePack inference code.

### CODE REVIEW

**Comment by FurkanGozukara** [comment]
> `--fp8` and `--fp8_scaled` are supported. With 24GB VRAM, no block swap is required with these options, making it about 30% faster than the official FramePack repository.

wow amazing

i wish you had added fp8 to this repo too i failed to do : https://github.com/Tencent/InstantCharacter

**Comment by Sarania** [comment]
I've ported Kohya's fp8 scaled to several projects(HunyuanVideoWrapper, WanVideoWrapper, and my own local projects including LLMs). You need a copy of the /musubi-tuner/modules/fp8_optimization_utils.py file copied to your new project somewhere. Then import apply_fp8_monkey_patch, optimize_state_dict_with_fp8 and at the point the full fp16/bf16 model is loaded to the cpu(which you can make it be if necessary) just do:
```
  state_dict = model.state_dict()
  state_dict = optimize_state_dict_with_fp8(state_dict, device, target_layer_keys=target_keys, exclude_layer_keys=exclude_keys, move_to_device=False)
  apply_fp8_monkey_patch(model, state_dict, False)  # True if you want fp8_fast
  model.load_state_dict(state_dict, strict=True, assign=True)
  ```
  
  model is your model, device is like "cuda" or such, target_layer_keys and exclude_keys are which keys you want to quantize and which you want to exclude and that varies per model. The algo already excludes any keys that don't end in "weight", you usually additionally wanna exclude "norm" and "head" plus any embeddings. That's all there is to it.
  
  Bonus: I have a little utility for loading and inspecting/converting models between dtypes. If you wanna inspect the key structure of your model to figure out what keys to include/exclude you can use it with --inspect which will print out all the keys, shapes, and dtypes.
  
  ```
  import os
import argparse
import torch
import safetensors
from safetensors.torch import save_file
from tqdm import tqdm

parser = argparse.ArgumentParser(description="Convert any model checkpoint to safetensors with dtype cast, or inspect it's keys.")
parser.add_argument("--input", required=True, help="Checkpoint to convert")
parser.add_argument("--output", help="Output safetensors file")
parser.add_argument("--verbose", action="store_true")
parser.add_argument("--convert", action="store_true")
parser.add_argument("--inspect", action="store_true")
parser.add_argument("--key_target", type=str)
parser.add_argument("--weights_only", type=str, default="true")
parser.add_argument("--dtype", type=str)
args = parser.parse_args()


def load_torch_file(ckpt, weights_only=True, device=None, return_metadata=False):
    if device is None:
        device = torch.device("cpu")
    metadata = None
    if ckpt.lower().endswith((".safetensors", ".sft")):
        try:
            with safetensors.safe_open(ckpt, framework="pt", device=device.type) as f:
                sd = {k: f.get_tensor(k) for k in f.keys()}
                metadata = f.metadata() if return_metadata else None
        except Exception as e:
            raise ValueError(f"Safetensors load failed: {e}\nFile: {ckpt}")
    else:
        pl_sd = torch.load(ckpt, map_location=device, weights_only=weights_only)
        sd = pl_sd.get("state_dict", pl_sd)
    return (sd, metadata) if return_metadata else sd


print("Loading checkpoint...")
weights_only = args.weights_only.lower() == "true"
checkpoint = load_torch_file(args.input, weights_only)

dtype_mapping = {
    "fp16": torch.float16,
    "float16": torch.float16,
    "bf16": torch.bfloat16,
    "bfloat16": torch.bfloat16,
    "fp32": torch.float32,
    "float32": torch.float32,
}

if args.convert and os.path.exists(args.output):
    confirm = input(f"{args.output} exists. Overwrite? [y/N]: ").strip().lower()
    if confirm != "y":
        print("Aborting.")
        exit()

converted_state_dict = {}
keys_to_process = (
    [k for k in checkpoint if args.key_target in k] if args.key_target else checkpoint.keys()
)

for key in tqdm(keys_to_process, desc="Processing tensors"):
    value = checkpoint[key]

    if args.verbose or args.inspect:
        print(f"{key}: {value.shape} ({value.dtype})")

    if args.convert:
        dtype_to_use = dtype_mapping.get(args.dtype.lower(), value.dtype) if args.dtype else value.dtype
        converted_state_dict[key] = value.to(dtype_to_use)

if args.convert:
    print(f"Saving converted tensors to '{args.output} {dtype_to_use}'...")
    save_file(converted_state_dict, args.output)
```

File: hv_train_network.py [hunk 1]
```diff
@@ -1049,6 +1053,8 @@ def handle_model_specific_args(self, args: argparse.Namespace):
 
         self._control_training = False  # HunyuanVideo does not support control training yet
 
+        self.default_guidance_scale = 6.0
```
[the remark this answers is not in the document]
Comment (kohya-ss) [reply hv_train_network.py:1056]: `hv_train_network.py` is for HunyuanVideo, so 6.0 is fine.

**Comment by maybleMyers** [comment]
I have made a version with support for first/last frames here: https://github.com/maybleMyers/H1111/blob/kohya-framepack/fpack_generate_video.py

**Comment by maybleMyers** [comment]
Also uploaded all files in one place here: https://huggingface.co/maybleMyers/framepack_h1111

**Comment by sdbds** [comment]
The current cache method seems to have an issue, as when I tried to cache a video file using the document command, there was no latent generation

cache latent failed, cache TE works.

```
[general]
resolution = [832,480]
caption_extension = ".txt"
batch_size = 1
enable_bucket = true
bucket_no_upscale = false
num_repeats = 1

[[datasets]]
video_directory = "./train/video"
cache_directory = "./train/video/cache"
target_frames = [1, 13, 25]
frame_sample = 4
frame_extraction = "full"
max_frames = 81
```

**Comment by sdbds** [comment]
> > The current cache method seems to have an issue, as when I tried to cache a video file using the document command, there was no latent generation
> 
> Could you please share the log for the dataset config? It will be something like this:
> 
> ```
> [Dataset 0]
>   is_image_dataset: False
>   resolution: (832, 480)
>   batch_size: 1
>   num_repeats: 1
>   caption_extension: ".txt"
>   enable_bucket: True
>   bucket_no_upscale: False
>   cache_directory: "..."
>   debug_dataset: False
>     video_directory: "..."
>     video_jsonl_file: "None"
>     control_directory: "None"
>     target_frames: (1, 13, 25)
>     frame_extraction: full
>     frame_stride: 1
>     frame_sample: 4
>     max_frames: 81
>     source_fps: None
> ```

```
INFO:__main__:Load dataset config from ./toml/qinglong-video-datasets.toml
INFO:dataset.image_video_dataset:glob videos in ./train/video
INFO:dataset.image_video_dataset:found 1 videos
INFO:dataset.config_utils:[Dataset 0]
  is_image_dataset: False
  resolution: (832, 480)
  batch_size: 1
  num_repeats: 1
  caption_extension: ".txt"
  enable_bucket: True
  bucket_no_upscale: False
  cache_directory: "./train/video/cache"
  debug_dataset: False
    video_directory: "./train/video"
    video_jsonl_file: "None"
    control_directory: "None"
    target_frames: (1, 13, 25)
    frame_extraction: full
    frame_stride: 1
    frame_sample: 4
    max_frames: 81
    source_fps: None


INFO:__main__:Loading VAE model from ./ckpts/framepack/hunyuan_video_vae_fp32.safetensors
INFO:hunyuan_model.vae:Loading 3D VAE model (884-16c-hy) from: ./ckpts/framepack/hunyuan_video_vae_fp32.safetensors
INFO:hunyuan_model.vae:VAE to dtype: torch.float16
INFO:frame_pack.framepack_utils:Set chunk_size to 32 for CausalConv3d
INFO:frame_pack.framepack_utils:Enabled spatial tiling with min size 256
INFO:__main__:Loading image encoder from ./ckpts/framepack/sigclip_vision_patch14_384.safetensors
INFO:frame_pack.framepack_utils:Loading image encoder feature extractor
INFO:frame_pack.framepack_utils:Loading image encoder from ./ckpts/framepack/sigclip_vision_patch14_384.safetensors
INFO:__main__:Encoding dataset [0]
1it [00:00,  1.56it/s]
```

**Comment by sdbds** [comment]
> > 1it [00:00,  1.56it/s]
> 
> Since this log is displayed, it seems that caching is taking place. It may have been removed when unnecessary files deletion in post-processing.
> 
> Is there a log such as `Removed old cache file` after this?

It seems to be due to the default enabled skip_existing, as previously skip_existing would cause issues with the te cache, so I only disabled the te part.
After I disabled it, it seems to be normal.
It's amazing, even on the first generation, enabling skip_existing can cause the generation to disappear.

**Comment by maybleMyers** [comment]
I get about 14.6s/it for a 640 bucket video during inference with no optimizations and 26 blocks swapped on a 4090 mobile (about 3090 speed w/ 16GB vram), are you swapping enough blocks?

I can confirm the caching works well and the training does not seem to be doing bad: steps:   0%|   | 17/14400 [10:06<142:33:35, 35.68s/it, avr_loss=0.0448] ((this is a normal 3090 with 848x480 buckets, [37, 73, 109] target frames.))

**Comment by sdbds** [comment]
> For inference, I get this with RTX 4090, 832x480, no block swap, with `--fp8_scaled`: 25/25 [01:17<00:00, 3.11s/it]
> 
> There may be some edge case problem, so if it doesn't improve, please share the command line options.

Now it seems to be working normally, not sure if it was due to GPU memory leak or other reasons...

At least for the moment, it seems that the entire set from caching, training, and inference has not encountered any other issues.
```
Extended arguments:
  --fp8_scaled
  --attn_mode=sageattn
  --output_type=both
  --seed=1026
  --fps=24
  --image_encoder=./ckpts/framepack/sigclip_vision_patch14_384.safetensors
  --text_encoder1=./ckpts/text_encoder/llava_llama3_fp16.safetensors
  --text_encoder2=./ckpts/text_encoder_2/clip_l.safetensors
  --vae_chunk_size=32
  --vae_spatial_tile_sample_min_size=128
  --image_path=toml/1.jpg
  --lora_weight
  ./output_dir/framepack_qinglong.safetensors
  --lora_multiplier
  1.0
  --prompt="Ichika Anon from BanG Dream! MyGO!!!!!, an anime girl with long, straight pink hair cascading past her shoulders and bright blue eyes, stands bathed in stage lighting. She's holding a light seafoam green electric guitar, which looks like an ESP model (the logo 'ESP' is visible on the black strap and the headstock). She wears a stylish blue zip-up jacket with ruffled details over a dark top, paired with a short white pleated skirt and a simple black choker. A confident, slightly playful smile is on her face as she looks directly forward. The camera angle is slightly low, emphasizing her presence on what appears to be a dimly lit stage, with strong spotlights creating lens flares and highlighting her figure against the dark background. She lightly strums the guitar as the shot holds.
"
  --video_size
  832
  480
  --infer_steps=25
```

**Comment by sdbds** [comment]
> I'm working on specifying the starting image for each section. I have committed the draft code (no documentation yet).
> 
> In this implementation, generate each section as if it were the starting of a video.
> 
>  20250425-222816_166850308_.-.mp4 
> f1 - コピー
> 
> The arguments are like this:
> 
> > python fpack_generate_video.py --video_size 832 480 --video_seconds 5 --infer_steps 25 --prompt "0:A high school anime girl with short black hair, glasses, and striking red eyes, wearing a sailor uniform with a green neckerchief, turning around and smiles at the viewer while adjusting her glasses. After that, she turns red with embarrassment and waves her hands vigorously with a surprised look on her face. The camera perspective is from a standing position, as if the viewer is her classmate. Her blue hair highlights catching the light.;;;1:A high school anime girl with short black hair, glasses, and striking red eyes, wearing a sailor uniform with a green neckerchief, smiling at the viewer while adjusting her glasses. After that, she turns red with embarrassment and waves her hands widely with a surprised face. The camera perspective is from a standing position, as if the viewer is her classmate. Her blue hair highlights catching the light.;;;2:A high school anime girl with short black hair, glasses, and striking red eyes, wearing a sailor uniform with a green neckerchief, blushing with embarrassment. And she waves her hands widely with a surprised face. The camera perspective is from a standing position, as if the viewer is her classmate. Her blue hair highlights catching the light.;;;3:A high school anime girl with short black hair, glasses, and striking red eyes, wearing a sailor uniform with a green neckerchief, moving her arms vigorously with a deformed surprised face. The camera perspective is from a standing position, as if the viewer is her classmate. Her blue hair highlights catching the light."　--image_path "0:..\images\sec0.png;;;1:..\images\sec1.png;;;2:..\images\sec2.png;;;3:..\images\sec3.png" ...
> 
> Note that the prompts for each section describe what happens from there to the end.

I think this feature is very suitable for GIFs or webp animated images.
Can we use the interpolation frame method to expand the animated image into a video, or add gif, webp training?

**Comment by Sarania** [comment]
You mean frame interpolation? One way is GIMM-VFI, which I use for Wan video outputs in general:

https://github.com/Sarania/musubi-tuner/blob/add-features/blissful_tuner/GIMMVFI.py

I VFI all my Wan outptus from 16 -> 32 FPS.

```
python GIMMVFI.py --input videofile --model /path/to/gimm-vfi/ --dtype fp16
```

Models: https://www.dropbox.com/scl/fi/tcq68jxr52o2gi47eup37/gimm-vfi.7z?rlkey=skvzwxi9lv9455py5wrxv6r5j&st=ho4e3hz1&dl=0

It's really good and capable of up to 8x increase though higher rates increase artifacts(2x basically has none though!) This example spits out ProRes in an MKV because that's what I use to avoid generational loss, but you can easily change it to mp4 or w/e it's just writing out with ffmpeg through a common video processing class in https://github.com/Sarania/musubi-tuner/blob/add-features/blissful_tuner/video_processing_common.py that's also used by my upscaler and face fix scripts. Anyway I don't know if it's helpful for what you're after but feel free to make use of it or not as usual 🙂

**Comment by maybleMyers** [comment]
Thanks kohya. I tried specifying the starting image for each section and it works great! I can also confirm lora training is working well after a day or so of letting it run. This was the launch command:
accelerate launch --num_cpu_threads_per_process 2 --mixed_precision bf16 fpack_train_network.py \
    --dit hunyuan/FramePackI2V_HY_bf16.safetensors\
    --vae hunyuan/pytorch_model.pt \
    --text_encoder1 hunyuan/llava_llama3_fp16.safetensors \
    --text_encoder2 hunyuan/clip_l.safetensors \
    --image_encoder hunyuan/model.safetensors \
    --dataset_config model/fpbros/fpbros.toml \
    --sdpa --mixed_precision bf16 \
    --optimizer_type adamw8bit --learning_rate 2e-4 --gradient_checkpointing \
    --timestep_sampling shift --weighting_scheme none --discrete_flow_shift 3.0 \
    --max_data_loader_n_workers 2 --persistent_data_loader_workers \
    --network_module networks.lora_framepack --network_dim 32 \
    --max_train_epochs 100 --save_every_n_epochs 1 --seed 42 \
    --output_dir model/fptest/model --output_name fptest \
    --blocks_to_swap 20
